From 75bc2dd40d86ca2d8e647a907080009943805cf2 Mon Sep 17 00:00:00 2001
From: Jonathan Bell <jonathan@raspberrypi.com>
Date: Wed, 15 Sep 2021 17:56:45 +0100
Subject: [PATCH] sound/usb: call usb_autopm_get_interface() for devices that
 should not be suspended

Webcams with microphones are composite devices, and autosuspend is set
at the device level. If uvcvideo is probed after snd-usb-audio, the effect
of the quirk applied by snd-usb-audio is undone by uvcvideo's global
application of autosuspend.

Incrementing the interface's PM refcount in such cases prevents runtime PM
from happening, thus the device is left active.

Signed-off-by: Jonathan Bell <jonathan@raspberrypi.com>
---
 sound/usb/card.c | 8 +++++++-
 1 file changed, 7 insertions(+), 1 deletion(-)

diff --git a/sound/usb/card.c b/sound/usb/card.c
index 1764b9302d46..fbbe330a52e3 100644
--- a/sound/usb/card.c
+++ b/sound/usb/card.c
@@ -817,8 +817,14 @@ static int usb_audio_probe(struct usb_interface *intf,
 	if (ignore_ctl_error)
 		chip->quirk_flags |= QUIRK_FLAG_IGNORE_CTL_ERROR;
 
-	if (chip->quirk_flags & QUIRK_FLAG_DISABLE_AUTOSUSPEND)
+	if (chip->quirk_flags & QUIRK_FLAG_DISABLE_AUTOSUSPEND) {
+		/*
+		* Grab the interface, because on a webcam uvcvideo may race
+		* with snd-usb-audio during probe and re-enable autosuspend.
+		*/
+		usb_autopm_get_interface(intf);
 		usb_disable_autosuspend(interface_to_usbdev(intf));
+	}
 
 	/*
 	 * For devices with more than one control interface, we assume the
-- 
2.32.0

